### PR TITLE
(V2.2 2/3) update mod form to include grade attributes inputs

### PR DESCRIPTION
**Pull request by octaviorequena**
Source: [NTGR-668](https://accredible.atlassian.net/browse/NTGR-668)

Technical Spec: [Moodle V2.2](https://accredible.atlassian.net/wiki/spaces/PD/pages/3147268170/Moodle+V2+Tech+Spec+-+Grade+Field?focusedCommentId=3186556949#UI---Form)

Video with new UI
https://user-images.githubusercontent.com/91325424/171600424-17f9d323-3704-429a-b675-d174320e24cc.mov

### CODE REVIEW

File: mod_form.php [hunk 1]
```diff
@@ -100,16 +102,27 @@ public function definition() {
...
-        if ($quizes = $DB->get_records_select('quiz', 'course = :course_id', array('course_id' => $id) )) {
+        $quizchoices = array(0 => 'Select a Quiz');
+        if ($quizes = $DB->get_records_select('quiz', 'course = :course_id', array('course_id' => $id), '', 'id, name')) {
             foreach ($quizes as $quiz) {
                 $quizchoices[$quiz->id] = $quiz->name;
             }
         }
 
+        // Load course assigments.
+        $assigmentschoices = array(0 => 'Select an Activity Grade');
```
Comment (kei178) [inline mod_form.php:113]: Question - Would't it be better to use `array('' => 'Select an Activity Grade');` instead of `0`?

File: mod_form.php [hunk 2]
```diff
@@ -138,6 +151,35 @@ public function definition() {
...
+            $attributekeyschoices = array('' => 'Select a Custom Attribute') + $attributekeys;
+            // Hidden element to check if we should disable the "gradeattributekeyname" select.
+            $mform->addElement('hidden', 'attributekysnumber', 1);
+        } else {
+            $mform->addElement('hidden', 'attributekysnumber', 0);
+        }
+
+        $mform->addElement('checkbox', 'includegradeattribute', get_string('includegradeattributedescription', 'accredible'),
+            get_string('includegradeattributecheckbox', 'accredible'));
+        if (isset( $accrediblecertificate->includegradeattribute )) {
```
Comment (kei178) [inline mod_form.php:166]: When adding a new Accredible module, the grade fields are not hidden even though the checkbox is not checked.

Screen Shot 2022-06-03 at 14 13 39

File: mod_form.php [hunk 3]
```diff
@@ -138,6 +151,35 @@ public function definition() {
...
+        $mform->addElement('html', $includegradewrapperhtml);
+        $mform->addElement('select', 'gradeattributegradeitemid', get_string('gradeattributegradeitemselect', 'accredible'),
+            $assigmentschoices, $inputstyle);
+        $mform->addElement('select', 'gradeattributekeyname', get_string('gradeattributekeynameselect', 'accredible'),
+            $attributekeyschoices, $inputstyle);
+        $mform->disabledIf('gradeattributekeyname', 'attributekysnumber', 'eq', 0);
+        $mform->addElement('static', 'emptygradeattributekeyname', '', get_string('emptygradeattributekeyname', 'accredible',
+            $dashboardurl));
+        $mform->addElement('html', '</div>');
+
```
Comment (kei178) [inline mod_form.php:182]: You also need to update `accredible_add_instance` and `accredible_update_instance` in `lib.php` to save the selected values for the DB record.

File: mod_form.php [hunk 4]
```diff
@@ -138,6 +151,35 @@ public function definition() {
             $mform->addElement('static', 'additionalactivitiestwo', '', get_string('additionalactivitiestwo', 'accredible'));
         }
 
+        // Load Accredible attribute keys.
+        $attributekeys = $attributekeysclient->get_attribute_keys();
```
Comment (kei178) [inline mod_form.php:155]: We should query the list of attribute keys with `kind: “text”`

**Review by kei178 — CHANGES_REQUESTED** [review]
@octaviorequena - I left several comments on the implementation and the behaivor.

File: mod_form.php (hunk 4 above)
Comment (octaviorequena) [reply mod_form.php:155]: I have added that parameter as a default one when requesting `attribute_keys`.

File: mod_form.php (hunk 3 above)
Comment (octaviorequena) [reply mod_form.php:182]: Added

File: lib.php [hunk 5]
```diff
@@ -228,26 +231,22 @@ function accredible_update_instance($post) {
...
+    $dbrecord->finalquiz = $post->finalquiz;
+    $dbrecord->includegradeattribute = $post->includegradeattribute;
+    $dbrecord->gradeattributegradeitemid = $post->gradeattributegradeitemid;
+    $dbrecord->gradeattributekeyname = $post->gradeattributekeyname;
+
     // Save record.
     if ($accrediblecertificate->achievementid) {
-        $dbrecord = new stdClass();
-        $dbrecord->id = $post->instance;
-        $dbrecord->achievementid = $post->achievementid;
```
Comment (kei178) [inline lib.php:235]: I think we should keep this.

**Review by kei178 — CHANGES_REQUESTED** [review]
@octaviorequena - It throws an error when the `includegradeattribute` checkbox is not checked.

**Review by kei178 — CHANGES_REQUESTED** [review]
@octaviorequena - I think the issue has not been fixed on the update action. Could you take another look?

File: mod_form.php [hunk 6]
```diff
@@ -111,17 +113,28 @@ public function definition() {
...
+        $quizchoices = array('' => 'Select a Quiz');
+        if ($quizes = $DB->get_records_select('quiz', 'course = :course_id', array('course_id' => $id), '', 'id, name')) {
             foreach ($quizes as $quiz) {
                 $quizchoices[$quiz->id] = $quiz->name;
             }
         }
 
+        // Load course assigments.
+        $assigmentschoices = array('' => 'Select an Activity Grade');
+        $assigments = $DB->get_records_select('grade_items', 'courseid = :course_id AND itemtype = :item_type',
```
Comment (kei178) [inline mod_form.php:125]: Minor - This variable will not include only `assignments` but also `quizzes`, `forums`, etc., any gradable modules.

**Review by naveenbalachandar — APPROVED** [review]

**Pull request merged**